src/pb_stub.cc

```diff
@@ -837,6 +862,130 @@ Stub::GetOrCreateInstance()
   return Stub::stub_instance_;
 }
 
+void
+Stub::LaunchLogRequestThread()
+{
+  log_thread_ = true;
+  log_monitor_ = std::thread(&Stub::ServiceLogRequests, this);
+}
+
+void
+Stub::TerminateLogRequestThread()
+{
+  log_thread_ = false;
+  log_monitor_.join();
+}
+
+void
+Stub::EnqueueLogRequest(PbLog* log_ptr)
+{
+  log_request_buffer.push(log_ptr);
+}
+
+void
+Stub::ServiceLogRequests()
+{
+  while (log_thread_) {
+    if (log_request_buffer.size() != 0) {
+      PbLog* log_request = log_request_buffer.front();
+      log_request_buffer.pop();
+      SendLogMessage(
+          log_request->Filename(), log_request->Line(), log_request->Message(),
+          log_request->Level(), log_request->Verbosity());
+      delete log_request;
+      log_request = nullptr;
+    }
+  }
+}
+
+void
+Stub::SendLogMessage(
+    const std::string& filename, uint32_t line, const std::string& message,
+    LogLevel level, uint32_t verbosity)
+{
+  std::shared_ptr<PbString> file_name = PbString::Create(shm_pool_, filename);
+  std::shared_ptr<PbString> log_message = PbString::Create(shm_pool_, message);
+  AllocatedSharedMemory<LogSendMessage> log_send_message =
+      shm_pool_->Construct<LogSendMessage>();
+
+  LogSendMessage* send_message_payload = log_send_message.data_.get();
+  new (&(send_message_payload->log_mu)) bi::interprocess_mutex;
+  new (&(send_message_payload->log_cv)) bi::interprocess_condition;
+
+  send_message_payload->filename = file_name->ShmHandle();
+  send_message_payload->line = line;
+  send_message_payload->logMsg = log_message->ShmHandle();
+  send_message_payload->level = level;
+  send_message_payload->verbosity = verbosity;
+  send_message_payload->waiting_on_stub = false;
+
+  std::unique_ptr<IPCMessage> log_request_msg =
+      IPCMessage::Create(shm_pool_, false /* inline_response */);
+  log_request_msg->Command() = PYTHONSTUB_LogRequest;
+  log_request_msg->Args() = log_send_message.handle_;
+   
+  ScopedDefer _([send_message_payload] {
+    {
+      bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->log_mu};
+      send_message_payload->waiting_on_stub = false;
+      send_message_payload->log_cv.notify_all();
+    }
+  });
+
+  {
+    // Send a message to be caught by the log monitor thread in python_be.cc
+    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->log_mu};
+    SendIPCLogMessage(log_request_msg);
+    while (!send_message_payload->waiting_on_stub) {
+      send_message_payload->log_cv.wait(guard);
+    }
+  }
+}
+
+void
+Logger::log(
```

Comment:
Is `__line__()` a thing in python? Or do users have to implement something to get the line number for each call?

If users have to implement something for each call, I think this could be done in some kind of function decorator or wrapper to do that for them for a better user experience.

---

I don't know how robust this is, but from quick googling, an example decorator might look like this. It can probably be done more cleanly if looked into further:
```python
import sys 
from inspect import currentframe, getframeinfo

def log_decorator(func):
  def wrapper(*args, **kwargs):
      caller_frame = currentframe().f_back
      caller_info = getframeinfo(caller_frame)
      caller_line_number = caller_info.lineno
      caller_filename = caller_info.filename
      print(f"{caller_filename}::{caller_line_number}", end=" ")
      return func(*args, **kwargs)

  return wrapper 

def log_full(filename, line, message, level, verbosity):
  print(f"{filename}::{line} [{level}] - {message}")

@log_decorator
def log_wrapped(message, level, verbosity):
  print(f"[{level}] - {message}")
  
if __name__ == "__main__":
  msg = "Hello"
  level = "INFO"
  verbosity = 0
  lineno = 28

  log_full(__file__, lineno, msg, level, verbosity) # depends on user to get their line number everytime / define it themself
  log_wrapped(msg, level, verbosity) # line number retrieved automatically without user intervention
```

Note these two functions achieve the same output in this example:
```
rmccormick@rmccormick-dt:~$ python3 test.py 
test.py::28 [INFO] - Hello
test.py::29 [INFO] - Hello
```

Reply:
I took a look into decorator implementation, but found that a simple interface function in triton_python_backend_utils should suffice:

`class level(Enum):
    INFO = 0,
    WARNINGS = 1,
    ERRORS = 2,
    VERBOSE = 3

def log(message, log_level=level.INFO, verbosity=0):
    caller_frame = currentframe().f_back
    caller_info = getframeinfo(caller_frame)
    caller_line_number = caller_info.lineno
    caller_filename = caller_info.filename
    if verbosity != 0 and log_level != level.VERBOSE:
        print(str(verbosity) + " " + str(log_level))
        print("Warning at " + caller_filename + " " + str(caller_line_number) + ": Verbosity set but log level is not VERBOSE")
        return
    logger = Logger()
    if(log_level == level.INFO):
        logger.log_info(caller_filename, caller_line_number, message)
    elif(log_level == level.WARNINGS):
        logger.log_warn(caller_filename, caller_line_number, message)
    elif(log_level == level.ERRORS):
        logger.log_error(caller_filename, caller_line_number, message)
    elif(log_level == level.VERBOSE):
        logger.log_verbose(caller_filename, caller_line_number, message, verbosity)`

With this interface function we can also a achieve a greater level of abstraction. Now instead of specifying one of multiple log functions and instantiating and object, the user simply has a single log interface function and we handle the rest internally: 

`pb_utils.log("Hello!", pb_utils.level.INFO, verbosity=0) # Okay
pb_utils.log(message="Verbose Msg!", verbosity=1) # Error, verbosity but not verbose level
pb_utils.log(message="Verbose Msg Lvl 2!", log_level=pb_utils.level.VERBOSE, verbosity=2) # okay
pb_utils.log("Hello!") # okay --> will default to INFO log
pb_utils.log(message="Hello!", log_level=pb_utils.level.ERRORS) #okay
pb_utils.log("Info Msg!", pb_utils.level.INFO) #okay
pb_utils.log("Warn Msg!", pb_utils.level.WARNINGS) #okay
pb_utils.log("Error Msg!", pb_utils.level.ERRORS) #okay`

Reply:
Can you attach some example logs

Reply:
Where'd `verbosity` go? Are we only supporting a single verbosity level for python backend?

Reply:
The verbosity option was removed. It will be added back if there is a request for it:
https://github.com/triton-inference-server/core/pull/101

Reply:
**Example Logs:**
These lines of code:
```
pb_utils.log("Info Msg!", pb_utils.Logger.INFO) 
pb_utils.log(message="Verbose Msg!", level=pb_utils.Logger.VERBOSE)
pb_utils.log(message="Error Msg!", level=pb_utils.Logger.ERRORS) 
pb_utils.log_info("Info Msg!") 
pb_utils.log_warn("Warn Msg!") 
pb_utils.log_error("Error Msg!")
pb_utils.log_verbose =("Verbose Msg!")
```
will appear in the logs like this (verbose appears too if set):
```
I0725 23:02:15.542552 167 model.py:54] Info Msg!
E0725 23:02:15.542828 167 model.py:56] Error Msg!
I0725 23:02:15.542989 167 model.py:57] Info Msg!
W0725 23:02:15.543126 167 model.py:58] Warn Msg!
E0725 23:02:15.543398 167 model.py:59] Error Msg!
```